clients/tabby-agent/src/lsp/ChatEditProvider.ts

```diff
@@ -249,121 +256,129 @@ export class ChatEditProvider {
-            closeTag = responseCommentTag?.[1];
-          }
-          if (!closeTag) {
-            break;
-          }
-          const reg = this.createCloseTagMatcher(closeTag);
-          const match = reg.exec(edit.buffer);
-          if (!match) {
-            if (inTag === "document") {
-              edit.editedText += edit.buffer;
-            } else if (inTag === "comment") {
-              edit.comments += edit.buffer;
-            }
-            edit.buffer = "";
-          } else {
-            if (inTag === "document") {
-              edit.editedText += edit.buffer.substring(0, match.index);
-            } else if (inTag === "comment") {
-              edit.comments += edit.buffer.substring(0, match.index);
-            }
-            edit.buffer = edit.buffer.substring(match.index);
-            if (match[0] === closeTag) {
-              inTag = false;
-            }
-          }
-          changed = true;
-        }
-        if (changed) {
-          const editedLines = this.generateChangesPreview(edit);
-          await this.connection.workspace.applyEdit({
-            edit: {
-              changes: {
-                [edit.location.uri]: [
-                  {
-                    range: edit.editedRange,
-                    newText: editedLines.join("\n") + "\n",
-                  },
-                ],
-              },
-            },
-          });
-          edit.editedRange = {
-            start: {
-              line: edit.editedRange.start.line,
-              character: 0,
-            },
-            end: {
-              line: edit.editedRange.start.line + editedLines.length,
-              character: 0,
-            },
-          };
+          const closeTag = inTag === "document" ? responseDocumentTag[1] : responseCommentTag?.[1];
+          if (!closeTag) break;
+
+          inTag = processBuffer(edit, inTag, closeTag);
+
+          await applyEdit(edit, isFirstEdit, false);
+          isFirstEdit = false;
         }
       }
+
+      if (this.currentEdit) {
+        this.currentEdit.state = "completed";
+        await applyEdit(this.currentEdit, false, true);
+      }
     } catch (error) {
-      await finalize("stopped");
-      // FIXME(@icycodes): use openai for nodejs instead of tabby-openapi schema
-      if (error instanceof TypeError && error.message.startsWith("terminated")) {
-        // ignore server side close error
-      } else {
+      if (this.currentEdit) {
+        this.currentEdit.state = "stopped";
+        await applyEdit(this.currentEdit, false, true);
+      }
+      if (!(error instanceof TypeError && error.message.startsWith("terminated"))) {
         throw error;
       }
+    } finally {
+      this.currentEdit = null;
+      this.mutexAbortController = null;
+    }
+  }
+
+  private async applyWorkspaceEdit(params: ApplyWorkspaceEditParams): Promise<boolean> {
+    try {
+      const result = await this.connection.sendRequest(ApplyWorkspaceEditRequest.type, params);
```

Comment:
We can add a client capabilities flag to indicate if the client supports this method, rather than using a try-catch fallback. However, it is acceptable for now.   
I believe this improvement could be a separate PR after this PR is merged.

Reply:
sorry for late response, I'll put this PR on my calendar